src/connection.ts

```diff
@@ -1,71 +1,184 @@
-	connection.onRequest('workspace/xreferences', (params, token) => handler.getWorkspaceReference(params, token));
-	connection.onRequest('workspace/xpackages', (params, token) => handler.getPackages(params, token));
-	connection.onRequest('workspace/xdependencies', (params, token) => handler.getDependencies(params, token));
+export function registerLanguageHandler(
+	messageEmitter: MessageEmitter,
+	messageWriter: MessageWriter,
+	handler: TypeScriptService,
+	logger: Logger = new NoopLogger(),
+	tracer = new Tracer()
+): void {
+	/** Tracks Subscriptions for results to unsubscribe them on $/cancelRequest */
+	const subscriptions = new Map<string | number, Subscription>();
+	messageEmitter.on('message', async message => {
+		logger.log('-->', message);
+		// Ignore responses
+		if (!isRequestMessage(message) && !isNotificationMessage(message)) {
+			return;
+		}
+		if (message.method === '$/cancelRequest' && isNotificationMessage(message)) {
+			// Cancel another request by unsubscribing from the Observable
+			const subscription = subscriptions.get(message.params.id);
+			if (!subscription) {
+				logger.error(`$/cancelRequest for unknown request ID ${message.params.id}`);
+				return;
+			}
+			subscription.unsubscribe();
+			subscriptions.delete(message.params.id);
+			return;
+		}
+		// If message is request and has tracing metadata, extract the span context and create a span for the method call
+		let span = new Span();
+		if (hasMeta(message) && isRequestMessage(message)) {
+			const context = tracer.extract(FORMAT_TEXT_MAP, message.meta);
+			if (context) {
+				span = tracer.startSpan('Handle ' + message.method, { childOf: context });
+				span.setTag('params', inspect(message.params));
+			}
+		}
+		const method = camelCase(message.method);
+		if (isRequestMessage(message) && typeof (handler as any)[method] !== 'function') {
+			// Method not implemented
+			messageWriter.write({
+				jsonrpc: '2.0',
+				id: message.id,
+				error: {
+					code: ErrorCodes.MethodNotFound,
+					message: `Method ${method} not implemented`
+				}
+			} as ResponseMessage);
+		}
+		// Call handler method with params and span
+		const returnValue = (handler as any)[method](message.params, span);
+		// Convert return value to Observable that emits a single item, the result (or an error)
+		let observable: Observable<any>;
+		if (returnValue instanceof Observable) {
+			observable = returnValue.take(1);
+		} else if (isPromiseLike(returnValue)) {
+			// Convert Promise to Observable
+			observable = Observable.from(returnValue);
+		} else {
+			// Convert synchronous value to Observable
+			observable = Observable.of(returnValue);
+		}
+		if (isRequestMessage(message)) {
+			// If request, subscribe to result and send a response
+			const subscription = observable
+				.finally(() => {
+					// Finish span
+					span.finish();
+					// Delete subscription from Map
+					subscriptions.delete(message.id);
```

Comment:
is it possible that this could run before `subscriptions.set(message.id, subscription);` below?